JP-3930: Make sure input is not modified inside Steps

melanieclarke opened the pull request:
<!-- If this PR closes a JIRA ticket, make sure the title starts with the JIRA issue number,
for example JP-1234: <Fix a bug> -->
Resolves [JP-3930](https://jira.stsci.edu/browse/JP-3930)

<!-- If this PR closes a GitHub issue that is not attached to a Jira ticket, uncomment the line below, and reference it here by its number -->
<!-- Closes # -->

<!-- describe the changes comprising this PR here -->
Adding smoke tests for all steps for as many input cases as possible and fixing any issues with input modification revealed by the tests.

Some steps addressed separately:
background: #9688
coron: #9693 
outlier_detection: #9709
skymatch, tweakreg, assign_mtwcs: #9723
tso_photometry: #9579 and/or #9627

<!-- if you can't perform these tasks due to permissions, please ask a maintainer to do them -->
## Tasks
- [ ] If you have a specific reviewer in mind, tag them.
- [x] add a build milestone, i.e. `Build 12.0` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [x] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [x] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see [changelog readme](https://github.com/spacetelescope/jwst/blob/main/changes/README.rst) for instructions) 
  - [x] update or add relevant tests
  - [ ] update relevant docstrings and / or `docs/` page
  - [x] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

melanieclarke commented on jwst/assign_mtwcs/assign_mtwcs_step.py:
```diff
@@ -38,9 +38,13 @@ def process(self, input_lib):
                 log.warning("Input data type is not supported.")
-                record_step_status(input_lib, "assign_mtwcs", False)
+                log.debug(f"Error was: {err}")
+                try:
+                    record_step_status(input_lib, "assign_mtwcs", False)
+                except AttributeError:
+                    pass
```
Small change here to avoid a confusing error message for types that couldn't be opened as a library.

melanieclarke commented on jwst/extract_2d/extract_2d.py:
```diff
@@ -106,10 +104,12 @@ def extract2d(
 
     # Set the step status to COMPLETE
-    output_model.meta.cal_step.extract_2d = "COMPLETE"
+    if output_model.meta.cal_step.extract_2d != "SKIPPED":
+        output_model.meta.cal_step.extract_2d = "COMPLETE"
```
Fix for case where the `nirspec` module set the status to SKIPPED, but the top level module overrode it.

melanieclarke commented on jwst/gain_scale/gain_scale_step.py:
```diff
@@ -34,29 +34,30 @@ def process(self, step_input):
             # Is the gain_factor already populated in the input model?
-            if input_model.meta.exposure.gain_factor is None:
+            if result.meta.exposure.gain_factor is None:
                 # Try to get the gain factor from the gain reference file
-                self.gain_filename = self.get_reference_file(input_model, "gain")
-                gain_model = datamodels.GainModel(self.gain_filename)
+                gain_filename = self.get_reference_file(result, "gain")
+                if gain_filename != "N/A":
+                    self.log.info("Using GAIN reference file: %s", gain_filename)
+                    with datamodels.GainModel(gain_filename) as gain_model:
+                        gain_factor = gain_model.meta.exposure.gain_factor
+                else:
+                    gain_factor = None
```
Small reorganization here to allow the gain model to be gracefully overridden

melanieclarke commented on jwst/guider_cds/guider_cds_step.py:
```diff
@@ -18,6 +18,8 @@ class GuiderCdsStep(Step):
 
     class_alias = "guider_cds"
 
+    reference_file_types = ["gain", "readnoise"]
```
Adding `reference_file_types` to the step so they can be overridden

melanieclarke commented on jwst/persistence/persistence_step.py:
```diff
@@ -102,8 +102,9 @@ def process(self, step_input):
 
             if output_pers is not None:  # output file of persistence
-                self.save_model(output_pers, suffix="output_pers")
+                self.save_model(output_pers, suffix="output_pers", force=self.save_persistence)
```
Adding `force` to make sure the intermediate file is saved, even if save_results is False

melanieclarke commented on jwst/photom/photom_step.py:
```diff
@@ -38,73 +38,74 @@ def process(self, input_data):
-        try:
-            # Do the correction
-            phot = photom.DataSet(
+        with datamodels.open(input_data) as input_model:
```
Moved to `with datamodels.open` to make sure `input_model` gets closed.

melanieclarke commented on jwst/photom/photom_step.py:
```diff
@@ -38,73 +38,74 @@ def process(self, input_data):
+                self.log.warning("Photom step will be skipped")
+                result = input_model.copy()
+                result.meta.cal_step.photom = "SKIPPED"
+                return result
```
Added a `return` here so that processing is not attempted for unsupported types

melanieclarke commented on jwst/pixel_replace/pixel_replace_step.py:
```diff
@@ -91,19 +91,17 @@ def process(self, input_data):
                         | datamodels.CubeModel,
                     ):
                         self.log.debug(f"Input is a {str(model)}.")
+                        replacement = PixelReplacement(model, **pars)
+                        replacement.replace()
+                        output_model[i] = replacement.output
+                        success = True
```
Slight reorganization here for simplicity - the `run_pixel_replace` flag is unnecessary.

melanieclarke commented on jwst/spectral_leak/spectral_leak_step.py:
```diff
@@ -116,8 +118,9 @@ def process(self, input_data):
-                    "Data sent to spectral_leak step is not a ModelContainer."
-                    "It is {type(input_model)}."
+                    "Data sent to spectral_leak step is not a ModelContainer. "
+                    f"It is {type(input_model)}."
                 )
+                result.meta.cal_step.spectral_leak = "SKIPPED"
```
Fix log string and add "SKIPPED" status

melanieclarke commented on jwst/srctype/srctype.py:
```diff
@@ -63,14 +63,9 @@ def set_source_type(input_model, source_type=None):
         # backwards compatibility with old datasets by first looking for the
         # SRCTYAPT keyword and using it if available, and if not, then use
         # SRCTYPE as both input and output (as before).
-        try:
```
Reorganize to remove some unreachable code

melanieclarke commented on jwst/srctype/srctype_step.py:
```diff
@@ -36,23 +36,13 @@ def process(self, step_input):
         """
-        if self.source_type is not None:
-            self.source_type = self.source_type.upper()
+        with datamodels.open(step_input) as input_model:
```
Reorganize to close the open `input_model` and remove some unnecessary code

melanieclarke commented on jwst/wavecorr/wavecorr.py:
```diff
@@ -44,15 +44,13 @@ def do_correction(input_model, wavecorr_file):
         input_model.meta.cal_step.wavecorr = "SKIPPED"
         return input_model
 
-    output_model = input_model.copy()
```
Moved the copy to the calling step instead of the algorithm

melanieclarke commented on jwst/wfss_contam/wfss_contam.py:
```diff
@@ -400,7 +400,13 @@ def contam_corr(
-    good_slits = [slit for slit in input_model.slits if slit.source_id in obs.source_ids]
+
+    # Copy over matching slits
+    good_slits = [
+        datamodels.SlitModel(slit.instance).copy()
+        for slit in input_model.slits
+        if slit.source_id in obs.source_ids
+    ]
```
Copying slits instead of making a reference to them to avoid modifying input

melanieclarke reviewed (commented):
Noting all places where I made changes other than simple `.copy()` or test additions.

melanieclarke commented:
Experimenting with # 2 for detector1 only in a branch on my fork shows no peak memory gains for this approach:
https://github.com/spacetelescope/RegressionTests/actions/runs/16780459179

So, I think I am instead inclined to take approach # 3 and merge these changes to make sure steps don't modify their input, then look at approach # 1 to improve memory performance more substantially.

emolter reviewed (commented):
If I understand https://github.com/spacetelescope/jwst/pull/9723 correctly, the plan would be to eventually add a call to `prepare_output` to basically every step.  This would make copies when necessary, and that would make certain changes you've added here redundant.  I'm looking for example at the copy calls you're adding to `badpix_selfcal_step` and `charge_migration_step`.

What is the benefit of merging these changes now, if sometime in the next few weeks we're going to start updating everything to use `prepare_output` instead?  Won't we have to revert changes like those copies?

penaguerrero reviewed (approved):
Thanks for adding all this test, Melanie! I think this looks good.

emolter commented on jwst/ami/tests/test_ami_analyze.py:
```diff
@@ -46,16 +43,9 @@ def test_ami_analyze_step_no_affine(example_model):
     AmiAnalyzeStep.call(example_model, affine2d=None)
 
 
-def create_throughput(nelem):
```
whoops I should have deleted this a long time ago when I added all the conftest methods

emolter commented on jwst/ami/tests/test_ami_analyze.py:
```diff
@@ -46,16 +43,9 @@ def test_ami_analyze_step_no_affine(example_model):
+    results = AmiAnalyzeStep.call(example_model)
+    assert example_model.meta.cal_step.ami_analyze is None
+    for result in results:
+        assert result is not example_model
```
For all tests in this PR, is there any kind of check that `result` is a deep copy instead of a shallow copy?  I'm thinking of the behavior documented by https://github.com/spacetelescope/stdatamodels/issues/512 where `result is not input_model` would be True but they are actually sharing the same data array

emolter commented on jwst/assign_mtwcs/assign_mtwcs_step.py:
```diff
@@ -38,9 +38,13 @@ def process(self, input_lib):
             try:
                 input_lib = ModelLibrary(input_lib)
-            except Exception:
+            except Exception as err:
```
I know you didn't write this pattern initially, but what kind(s) of exception is being caught here?  Should this be more specific?

emolter commented on jwst/assign_mtwcs/tests/test_mtwcs.py:
```diff
@@ -70,3 +70,16 @@ def test_mt_slitmodel(errtype):
+    file_path = get_pkg_data_filename("data/test_mt_asn.json", package="jwst.assign_mtwcs.tests")
+    with datamodels.open(file_path) as model:
+        result = AssignMTWcsStep.call(model)
+        assert result is not model
```
why isn't `.meta.cal_step` checked here the way it is for other tests in this PR?

melanieclarke replied:
This step is partially addressed here, and partially in the `prepare_output` PR.  I can move the changes here to the other PR.

melanieclarke replied:
Moved changes to assign_mtwcs to #9723

emolter commented on jwst/assign_mtwcs/tests/test_mtwcs.py:
General question: it seems like this PR adds one or more calls to the entire Step for basically every step in the pipeline.  How much does that affect the overall runtime of the unit test suite?

emolter commented on jwst/assign_wcs/tests/test_miri.py:
```diff
@@ -181,6 +181,16 @@ def test_mrs_tso_bounding_box():
     assert_allclose(cube.meta.wcs.bounding_box, ((-0.5, 49.5), (-0.5, 39.5)))
 
 
+def test_output_is_not_input():
```
why is this one here, but the tests for all the other modes ended up in `test_assign_wcs_step.py`?

melanieclarke replied:
Looking back at it now, I'm not sure.  I think I added this one first, when I tackled the steps at the beginning of the alphabet, then came back for more tests later when I learned more about the various ways the input can be modified.  I can move it to `test_assign_wcs_step.py` with the others.

emolter commented on jwst/badpix_selfcal/tests/test_badpix_selfcal.py:
```diff
@@ -276,9 +295,11 @@ def test_apply_flags(background):
-    """Test the badpix_selfcal step. This is a functional test that checks
-    that the step runs without error. The output will be checked by the
-    regtest.
+    """
+    Test the badpix_selfcal step.
+
+    This is a functional test that checks that the step runs without error.
+    The output will be checked by the regtest.
     """
```
"""Smoke test for the badpix_selfcal step."""

emolter commented on jwst/dq_init/dq_init_step.py:
is there a separate bug that was fixed here, to do with what happens when invalid inputs are passed in?

emolter commented on jwst/flatfield/tests/test_flatfield.py:
```diff
@@ -76,35 +76,25 @@ def test_flatfield_step_interface(instrument, exptype):
     assert result.meta.cal_step.flat_field == "COMPLETE"
 
-
-def exptypes():
```
why is so much random stuff removed from this test, like attribute setting for `data.meta` and this entire function?

emolter commented:
Thanks for understanding.  If we decide the best thing to do is to go forward with this without breaking it up, I'm happy to look at changes to certain specific steps.  (Isn't there some kind of semi-official "step maintainers" page somewhere or did that get deleted for being out-of-date?)

melanieclarke commented:
> (Isn't there some kind of semi-official "step maintainers" page somewhere or did that get deleted for being out-of-date?)

I moved it to an innerspace page and updated it a few months ago.  I'll send you the link in slack.

melanieclarke commented on jwst/extract_1d/extract_1d_step.py:
```diff
@@ -122,6 +122,9 @@ def _extract_soss(self, model):
             The output spectra.
         """
+        # Work on a copy of the model
+        model = model.copy()
```
I think I put it here instead because it is closer to the place where it is needed (where the SKIPPED status is set and the original model is returned).

I can move it to process, before calling `_extract_soss` if you think that will be easier to spot?

tapastro reviewed (approved):
Looks good to me! I don't see any lingering questions. The last RT run was clean - I assume we are not expecting failures? We can get this in and start a new run to be safe.

melanieclarke commented:
> Looks good to me! I don't see any lingering questions. The last RT run was clean - I assume we are not expecting failures? We can get this in and start a new run to be safe.

No expected failures!  I'll go ahead and merge and start a new run.
